### ISSUE TITLE
[Question] The App_id in the input_token did not match the Viewing

### ISSUE THREAD

**Issue by farahjr** [firebase/quickstart-unity#1040]
* Unity editor version: 2021.1.4
  * Firebase Unity SDK version: 7.1.0
  * Source you installed the SDK: .unitypackage
  * Problematic Firebase Component: Auth
  * Other Firebase Components in use: Database
  * Additional SDKs you are using: Facebook
  * Platform you are using the Unity editor on: Android
  * Platform you are targeting: Any
  * Scripting Runtime: IL2CPP

When calling 

SignInWithCredentialAsync encountered an error: System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> Firebase.FirebaseException: 
The supplied auth credential is malformed or has expired. [ Unsuccessful debug_token response from Facebook: 
{"error":{"message":"(#100) 
The App_id in the input_token did not match the Viewing App","type":"OAuthException","code":100,"fbtrace_id":"AhvgYOYORtUeWHO3dGrDJeb"}} ]

![image](https://user-images.githubusercontent.com/27826334/117514926-66716100-af6b-11eb-8b5b-3ae1ec1f2686.png)
![image](https://user-images.githubusercontent.com/27826334/117515245-07f8b280-af6c-11eb-9a5f-ba107ce80ecb.png)
![image](https://user-images.githubusercontent.com/27826334/117515622-29a66980-af6d-11eb-8641-bafa16cc659b.png)
![image](https://user-images.githubusercontent.com/27826334/117515705-670af700-af6d-11eb-912c-85b89722677f.png)

With a valid facebook Token(tested at facebook token debugger).
With every ID on the right place.
On editor every thing works fine with debug token.
After Deploy e Run, i can't pass the Firebase.Auth SignInWithCredentialAsync.

Tested removing a digit from AppId in facebook configuration
![image](https://user-images.githubusercontent.com/27826334/117515932-0b8d3900-af6e-11eb-921a-9c3a0978ebea.png)

I have noticed my googleservice-info.plist and google-services.json son't have any mentions to facebook App id and i don't know if they should.

**Comment by paulinon** [comment]
Hi @farahjr,

It appears that this issue involves private information that's related to your project. You may get in touch with [Firebase Support](https://firebase.google.com/support) where the team can shed some more light on this issue.

With that, I'll be closing this ticket for now.

Thanks!

**Comment by farahjr** [comment]
> Hey @farahjr,
> 
> I found this [community post](https://stackoverflow.com/questions/65214756/facebook-login-using-firebase-results-in-invalid-oauth-access-token-error) that talks about an issue similar to yours. With that, could you verify if you've chosen "Build or Connect to a Game" in your Facebook Developer Console? This could be what's causing the issue.

Just realized this isn't the firebase sdk git-hub............sorry for inconvenience. DO You know where i can find it?

**Comment by paulinon** [comment]
Hi @farahjr, 

That repository doesn't exist at the moment, but you're at the right place for reporting possible SDK-related issues. In order to verify this, I'd like you to share a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of your project (make sure it's free from any PII) along with the complete steps you've taken before facing this issue.

**Comment by farahjr** [comment]
New Unity project with any 2021 versions.
Import latest Firebase SDK
Import latest Facebook SDK
Create an facebook app
Go to menu in UNity called Facebook => Configuration and set everything up with facebook app informations
Create an Firebase Auth and Firebase Database. 
Configure your firebase auth with facebook app informations.
Download the files Firebase will create and put them inside Assets folder.

Implemente FIrebase initialize:

```
    public class DB : MonoBehaviour
    {
        public static DB db;
        public static FirebaseApp app;
        public static DatabaseReference DBreference;
        void Awake()
        {
            FirebaseApp.CheckAndFixDependenciesAsync().ContinueWith(task => {
                var dependencyStatus = task.Result;
                if (dependencyStatus == DependencyStatus.Available)
                {
                    app = FirebaseApp.DefaultInstance;
                    InitializeFirebase();                    
                }
                else
                {
                    Debug.LogError(string.Format(
                      "Could not resolve all Firebase dependencies: {0}", dependencyStatus));
                }
            });

            DontDestroyOnLoad(gameObject);
            if (db == null)            
                db = this;            
        }
        private void InitializeFirebase()
        {
            DBreference = FirebaseDatabase.DefaultInstance.RootReference;
        }
    }    
```

Implement facebook initializer and login functions:

```
public class FbAuth : MonoBehaviour
    {        
        // Awake function from Unity's MonoBehavior
        void Awake()
        {
            if (!FB.IsInitialized)
            {
                Debug.Log("Initializing facebook App on Awake");
                FB.Init(InitCallback, OnHideUnity);
            }
            else
            {
                Debug.Log("Activating facebook App on Awake");
                FB.ActivateApp();
            }
        }
        private void InitCallback()
        {
            if (FB.IsInitialized)
            {
                if (FB.IsLoggedIn)
                {
                    Login();
                }
                else
                {
                    FB.ActivateApp();
                }
            }
            else
            {
                Debug.Log("Failed to Initialize the Facebook SDK");
            }
        }
        private void OnHideUnity(bool isGameShown)
        {
            if (!isGameShown)
            {
                // Pause the game - we will need to hide
                Time.timeScale = 0;
            }
            else
            {
                // Resume the game - we're getting focus again
                Time.timeScale = 1;
            }
        }
        public void Login()
        {
            var perms = new List<string>() { "gaming_profile", "email" };
            Debug.Log("Gettting acces Token");
            FB.LogInWithReadPermissions(perms, AuthCallback);
        }
        private void AuthCallback(ILoginResult result)
        {
            Debug.Log("This is your accessTOken: " + AccessToken.CurrentAccessToken);
            if (FB.IsLoggedIn)
                Authenticate(AccessToken.CurrentAccessToken.TokenString.Trim());    
            else
            {
                Debug.Log("User cancelled login");
            }
        }
        private void Authenticate(string accessToken)
        {
            FirebaseAuth auth = FirebaseAuth.DefaultInstance;
            Credential credential = FacebookAuthProvider.GetCredential(accessToken);
            if (auth != null)
            {
                Debug.Log("FireBaseAuth Credential:" + credential.ToString() + "|" + credential.Provider + " |Access Token: " + 
                    accessToken + "auth options" + auth.App.Options.ApiKey + " " +
                    auth.App.Options.AppId + " " +
                    auth.App.Options.ProjectId + " " +
                    auth.CurrentUser);
                auth.SignInWithCredentialAsync(credential).ContinueWith(task =>
                {
                    if (task.IsCanceled)
                    {
                        Debug.LogError("SignInWithCredentialAsync was canceled.");
                        return;
                    }
                    if (task.IsFaulted)
                    {
                        Debug.LogError(accessToken + "SignInWithCredentialAsync encountered an error: " + task.Exception);
                        return;
                    }

                    FirebaseUser newUser = task.Result;
                    Debug.LogFormat("User signed in successfully: {0} ({1})",
                        newUser.DisplayName, newUser.UserId);
                    GameControl.control.isLoggedIn = true;
                });
            }

            else
                Debug.Log("USER ALREADY LOGGED IN:" + auth.CurrentUser.UserId);
        }
    }
```

Create a login button and play buttons. Play will only get active when user is authenticated:
![image](https://user-images.githubusercontent.com/27826334/117978796-a12b1e80-b308-11eb-8f77-e9068686ccae.png)

Login button will use this function OnCLick:
![image](https://user-images.githubusercontent.com/27826334/117978886-b7d17580-b308-11eb-9582-04e951336461.png)

Attached both scripts into my main camera:
![image](https://user-images.githubusercontent.com/27826334/117979154-febf6b00-b308-11eb-9db1-a3173df66494.png)

You might get into some issues with deploy and everything. I have runned through many issues. I had to patch my fb sdk due to a class making legacy calls WWW instead of using other Unity methods, but i can manage to help configuring it.

Everything will work fine while in Editor. You will get and token from facebook website for debuging and firebase will validate it successfully with facebook.

Once you get to "deploy and play" You will have this problem: 

SignInWithCredentialAsync encountered an error: System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> Firebase.FirebaseException:
The supplied auth credential is malformed or has expired. [ Unsuccessful debug_token response from Facebook:
{"error":{"message":"(#100)
The App_id in the input_token did not match the Viewing App","type":"OAuthException","code":100,

Every on in the internet solved this with facebook configuration, but my configuration is already like theirs.

**Comment by paulinon** [comment]
Thanks for the additional input, @farahjr. While I try to reproduce this behavior, could you try choosing the "Consumer" type in creating your app and see if this made a difference?

**Comment by farahjr** [comment]
> Thanks for the additional input, @farahjr. While I try to reproduce this behavior, could you try choosing the "Consumer" type in creating your app and see if this made a difference?

This is already consumer type, I even created another app anemed consumer with consumer type to check it out, and iI think i might have discovered the issue.

The problem is on my previous screenshots:
![image](https://user-images.githubusercontent.com/27826334/118123389-933dd200-b3ca-11eb-923a-1ed440b90e16.png)

The first token is debug token supplied by Facebook for testing the App Runner 2 wich is a Consumer type app.
The second is the Token Facebook generated for my Account. It is pointing towards "Runner" witch is my first wrong attempt of creating an Facebook App. For some reason, when requested a token from my app, Facebook think it is Runner, but then, when Firebase tries to validate, it it is Runner 2, hence the Facebook returns an error: "The App_id in the input_token did not match the Viewing App".

Now i have deleted "Runner" and "consumer", and I'm running a full scan on my computer to check where "Runner" app_id could be hard coded in my computer. You can see from the evidences it is not used in my Facebook SDK configuration.

I have another clue, if You have many Apps and You set the hash Keys on all of them, Facebook might link Your request to the Wrong app, i think it uses the hash references you registered on the Apps, then it returns the first match of request by hash Id.

Once my full scan is finished, im gonna try to login again and check, then I will post it here.

**Comment by farahjr** [comment]
![image](https://user-images.githubusercontent.com/27826334/118132162-c76ac000-b3d5-11eb-923a-d2621fefbfd3.png)

Now the generated Access token have no App at all......trying to figure it out.

**Comment by farahjr** [comment]
I have solved the problem, but not so easily........
Just delete all, the facebook apps i had created and try to configure just one didn't solved.
I deleted both sdks and all related folders i could.
deleted all information online on both platforms i could.
reinstalled both sdks
patched with the facebook dll with http bug corrected(same as before).
configured everything on try and error.

Despite many errors being due configuration missing  I wasn't able to get the same error i was getting before and now the aplication is working. Thank You for helping motivating me to redo everything.....I was very tired phycally and mentally to do that.
